internal/impl/kafka/enterprise/schema_registry_output.go

```diff
@@ -183,3 +234,129 @@ func (o *schemaRegistryOutput) Close(_ context.Context) error {
+// configured Schema Registry output if present. Otherwise, it creates it, caches it and returns the generated ID.
+func (o *schemaRegistryOutput) getOrCreateSchemaID(ctx context.Context, ss franz_sr.SubjectSchema) (int, error) {
+	key := schemaLineageCacheKey{
+		id:        ss.ID,
+		versionID: ss.Version,
+		subject:   ss.Subject,
+	}
+	if destinationID, ok := o.schemaLineageCache.Load(key); ok {
+		return destinationID.(int), nil
+	}
+
+	if o.backfillDependencies {
+		if err := o.createSchemaDeps(ctx, ss, true); err != nil {
+			return 0, fmt.Errorf("failed to backfill dependencies for schema with subject %q and version %d: %s", ss.Subject, ss.Version, err)
+		}
+	}
+
+	return o.createSchema(ctx, key, ss)
+}
+
+// createSchemaDeps creates and caches all the dependencies of the current schema (both references and previous versions).
+func (o *schemaRegistryOutput) createSchemaDeps(ctx context.Context, ss franz_sr.SubjectSchema, backfillPrevVersions bool) error {
+	key := schemaLineageCacheKey{
+		id:        ss.ID,
+		versionID: ss.Version,
+		subject:   ss.Subject,
+	}
+	if _, ok := o.schemaLineageCache.Load(key); ok {
+		return nil
+	}
+
+	// Backfill references recursively.
+	for _, ref := range ss.References {
+		schema, err := o.inputClient.GetSchemaBySubjectAndVersion(ctx, ref.Subject, &ref.Version, false)
+		if err != nil {
+			return fmt.Errorf("failed to get schema for subject %q with version %d: %s", ref.Subject, ref.Version, err)
+		}
+
+		if err := o.createSchemaDeps(ctx, schema, true); err != nil {
+			return fmt.Errorf("failed to create schema dependencies: %s", err)
+		}
+	}
+
+	// Backfill previous schema versions in ascending order.
+	if ss.Version > 1 && backfillPrevVersions {
+		versions, err := o.inputClient.GetVersionsForSubject(ctx, ss.Subject, false)
+		if err != nil {
+			return fmt.Errorf("failed to get schema versions for subject %q: %s", ss.Subject, err)
+		}
+
+		slices.Sort(versions)
+		for _, version := range versions {
+			schema, err := o.inputClient.GetSchemaBySubjectAndVersion(ctx, ss.Subject, &version, false)
+			if err != nil {
+				return fmt.Errorf("failed to get schema for subject %q with version %d: %s", ss.Subject, version, err)
+			}
+
+			if err := o.createSchemaDeps(ctx, schema, false); err != nil {
+				return fmt.Errorf("failed to create schema dependencies: %s", err)
+			}
+		}
+	}
+
+	if _, err := o.createSchema(ctx, key, ss); err != nil {
+		return fmt.Errorf("failed to create schema: %s", err)
+	}
+
+	return nil
+}
+
+// createSchema creates and caches the provided schema.
+func (o *schemaRegistryOutput) createSchema(ctx context.Context, key schemaLineageCacheKey, ss franz_sr.SubjectSchema) (int, error) {
+	if destinationID, ok := o.schemaLineageCache.Load(key); ok {
+		return destinationID.(int), nil
+	}
+
+	// TODO: Use `CreateSchemaWithID()` when `translate_ids: false` after https://github.com/twmb/franz-go/pull/849
+	// is merged.
```

Comment:
Are we waiting on this before we merge?

Reply:
I think it's fine to leave it as a TODO for now. The main use case is to do the ID translation.